plugins/techdocs-backend/src/service/router.ts

```diff
@@ -179,15 +179,25 @@ export async function createRouter({
               return;
             }
           } else {
-            logger.info(
-              'Found pre-generated docs for this entity. Serving them.',
-            );
-            // TODO: re-trigger build for cache invalidation.
-            // Add build info in techdocs_metadata.json and compare it against
-            // the etag/commit in the repository.
-            // Without this, docs will not be re-built once they have been generated.
-            // Although it is unconventional that anyone will face this issue - because
-            // if you have an external storage, you should be using CI/CD to build and publish docs.
+            // Fetch the etag from stored metadata file
+            const { etag: storedEtag } = await publisher.fetchTechDocsMetadata({
+              namespace,
+              kind,
+              name,
+            });
+            // Get etag from preparer for comparison
+            const preparer = preparers.get(entity);
+            const { etag: preparerEtag } = await preparer.prepare(entity, {
+              etag: storedEtag,
+            });
```

Comment:
Take a look at https://github.com/backstage/backstage/blob/67cc1f0416a4caaaae5ec3d70d37a8faa6b550a2/plugins/techdocs-backend/src/DocsBuilder/builder.ts#L93-L108

The `prepare` call will throw a `NotModifiedError`. We should catch that and not re-build when the error is thrown. Otherwise build.

Thanks, and this is needed. :D 

Since the check unnecessarily happens on each request, the next step should be to maintain an in-memory cache where the check for "whether or not to build sites" should happen not more than once in 60 seconds or so.

Reply:
very nice, thanks!

If the in-memory cache part is a requirement for this PR to be accepted I'll get to it after the weekend :)

It seems to me that this type of cache should be implemented on the "Builder" level rather than at the routing level since it would be an equally important thing to have implemented for local builds or am I overthinking this?

Reply:
Yes, you are right that it should be on the builder level instead of the route. But if you look inside the `docsBuilder.build()`, you can see that the check is already being done there! I think we may be able to just do `docsBuilder.build()` in this block.

Reply:
Looking at that piece of code I'm not seeing any calls to `preparer.prepare()` before (at least there's no await calls to any of the `.get()` implementations meaning at best there would be a race-condition) the entity is being checked for metadata.

I believe that if the idea is to be able to just call `docsBuilder.build()` and let that handle all the cases it would need to be refactored into an async flow or at least adding an optional parameter `etag` used to compare the stored vs preparer version.